src/gui/maptools/qgsmaptoolshaperegistry.h

```diff
@@ -0,0 +1,104 @@
+/***************************************************************************
+                         qgsmaptoolshaperegistry.h
+                         ----------------------
+    begin                : January 2022
+    copyright            : (C) 2022 by Denis Rouzaud
+    email                : denis@opengis.ch
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSMAPTOOLSHAPEREGISTRY_H
+#define QGSMAPTOOLSHAPEREGISTRY_H
+
+#define SIP_NO_FILE
+
+#include "qgsabstractrelationeditorwidget.h"
+#include "qgsmaptoolshapeabstract.h"
+#include "qgis_gui.h"
+
+class QgsMapToolShapeMetadata;
+class QgsMapToolCapture;
+
+/**
+ * \ingroup gui
+ * \brief Keeps track of the registered shape map tools
+ * \since QGIS 3.24
+ */
+class GUI_EXPORT QgsMapToolShapeRegistry
+{
+    Q_GADGET
+  public:
+
+    /**
+     * Constructor
+     */
+    QgsMapToolShapeRegistry();
+
+    ~QgsMapToolShapeRegistry();
+
+    /**
+     * Adds a new shape map tool
+     */
+    void addMapTool( QgsMapToolShapeMetadata *mapTool SIP_TRANSFER );
+
+    /**
+     * Removes a registered relation widget with given \a widgetType
+     */
+    void removeMapTool( const QString &id );
+
+    //! Returns the list of map tools
+    QList<QgsMapToolShapeMetadata *> mapToolMetadatas() const {return mMapTools;}
+
+    //! Returns the map tool metadata for the gin \a id
+    QgsMapToolShapeMetadata *mapToolMetadata( const QString &id ) const;
+
+    /**
+     * Constructs the map tool at the given \a id for the given \a parentTool
```

Comment:
```suggestion
     * Constructs the map tool at the given \a id for the given \a parentTool
     * Caller takes ownership of the returned tool.
```

(although given that this isn't exposed to sip, we could potentially return a unique_ptr here. Or is exposing this to sip in your long term vision?)

Reply:
yes, that is the plan. Just waiting a bit to be sure about the API.